common/src/heuristics.rs

```diff
@@ -8,53 +12,241 @@ const SIGSEGV: u32 = 11;
+/// aware of when deploying these changes.
+///
+/// ## x86 and x86-64
+///
+/// x86 uses variable-width instruction encodings; subtract one byte from the return address to
+/// derive an address that should be within the calling instruction. This will provide an address
+/// within a calling instruction found directly prior to the return address.
+///
+/// ## ARMv6 and ARMv7
 ///
-/// See <https://goo.gl/g17EAn> for detailed information on this topic.
+/// - **Step 1:** Strip the low order thumb bit from the return address. ARM uses the low bit to
+///   inform the processor that it should enter thumb mode when jumping to the return address. Since
+///   all instructions are at least 2 byte aligned, an actual instruction address will never have
+///   the low bit set.
+///
+/// - **Step 2:** Subtract 2 Bytes. 32-bit ARM instructions are either 2 or 4 bytes long, depending
+///   on the use of thumb. This will place the symbolication address within the likely calling
+///   instruction. All ARM64 instructions are 4 bytes long; subtract 4 bytes from the return address
+///   to derive the likely address of the calling instruction.
+///
+/// # More Information
+///
+/// The above information was taken and slightly updated from the now-gone *PLCrashReporter Wiki*.
+/// An old copy can still be found in the [internet archive].
+///
+/// [internet archive]:
+/// https://web.archive.org/web/20161012225323/https://opensource.plausible.coop/wiki/display/PLCR/Automated+Crash+Report+Analysis
+///
+/// [`caller_address`]: struct.InstructionInfo.html#method.caller_address
+#[derive(Clone, Debug)]
 pub struct InstructionInfo {
-    /// The address of the instruction we want to use as a base.
+    #[doc(hidden)]
+    #[deprecated = "Use InstructionInfo::new() instead"]
     pub addr: u64,
-    /// The architecture we are dealing with.
+    #[doc(hidden)]
+    #[deprecated = "Use InstructionInfo::new() instead"]
     pub arch: Arch,
-    /// This is true if the frame is the cause of the crash.
+    #[doc(hidden)]
+    #[deprecated = "Use is_crashing_frame() instead"]
     pub crashing_frame: bool,
-    /// If a signal is known that triggers the crash, it can be stored here.
+    #[doc(hidden)]
+    #[deprecated = "Use signal() instead"]
     pub signal: Option<u32>,
-    /// The optional value of the IP register.
+    #[doc(hidden)]
+    #[deprecated = "Use ip_register_value() instead"]
     pub ip_reg: Option<u64>,
 }
 
 impl InstructionInfo {
+    /// Creates a new instruction info instance.
+    ///
+    /// By default, the frame is not marked as *crashing frame*. The signal and instruction pointer
+    /// register value are empty.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use symbolic_common::{Arch, InstructionInfo};
+    ///
+    /// let caller_address = InstructionInfo::new(Arch::X86, 0x1337)
+    ///     .caller_address();
+    /// ```
+    pub fn new(arch: Arch, instruction_address: u64) -> Self {
```

Comment:
When reading this in use `new` reads a little weird.  Though I'm struggling to come up with a better name, so I guess it's fine.

Reply:
Yeah, `new` is not very descriptive, but it seems to be the accepted name for such functions.